tinymce-single-react-ui/src/Turducken.js

```diff
@@ -0,0 +1,82 @@
+import BlockToolbar from './components/toolbar/BlockToolbar'
+import InlineToolbar from './components/toolbar/InlineToolbar'
+import TinyMCEReact from './components/tinymce/tinymce-react-ui'
+import {blockList, blockType, blockAlign, getTopLevelBlock} from './utils/tag'
+import '../../shared/post-content'
+import 'assets/stylesheets/main.scss'
+
+let blockOpen  = (collapsed) => (collapsed !== null &&  collapsed) // block  if caret
+let inlineOpen = (collapsed) => (collapsed !== null && !collapsed) // inline if range selection
+
+// get tiny node from the container, and the top level block from the caret node
+let tinyNode = (containerNode) => ((containerNode && containerNode.children.length > 0) ? containerNode.children[0] : null)
+let topLevelBlock = (tinyNode, node) => ((tinyNode && node) ? getTopLevelBlock(tinyNode, node) : null)
+
+// Rect at the start of the Range
+let findStartOfRange = (range) => {
+	// make a collapsed range at the start point
+	if (range) {
+		let r = range.cloneRange();
+		r.setEnd(range.startContainer, range.startOffset);
+		return r.getBoundingClientRect();
+	}
+}
+
+// Rect for the Range
+let rangeRect = (range) => {
+	if (range) {
+		return range.getBoundingClientRect();
+	}
+}
+
+export default function Turducken(props) {
+  let store = props.store
+  let collapsed = store.getState().get('collapsed')
+  let range = store.getState().get('range')
+  let node =  store.getState().get('node') // node of caret or ancestor of range
+  let editorRef = store.getState().get('editorRef')
+  let tiny = tinyNode(editorRef)
+  let topBlock = topLevelBlock(tiny, node)
+  let topRect = rangeRect(topBlock)
+
+  return (
+    <div>
+      { topRect && topRect.width &&
+        <div>
+          <div style={ {position: 'absolute', top: topRect.top + 'px',
+                                              width: topRect.width + 'px',
+                                              height: '2px', background: '#ccc' } }></div>
+
+          <div style={ {position: 'absolute', top: topRect.height + topRect.top + 'px', width: topRect.width + 'px', height: '2px', background: '#ccc' } }></div>
+
+          <div style={ {position: 'absolute', top: topRect.top + 'px', width: '2px', height: topRect.height + 'px', background: '#ccc' } }></div>
+
+          <div style={ {position: 'absolute', top: topRect.top + 'px', width: '2px', height: topRect.height + 'px', left: topRect.width + topRect.left + 'px', background: '#ccc' } }></div>
+
+        </div>
+      }
+      <div>
+        <InlineToolbar isOpen={ inlineOpen(collapsed) }
+          rangeRect={ findStartOfRange(range) }
+          pageYOffset={ window.pageYOffset }
+          node={ node }
+          />
+        <BlockToolbar  isOpen={ blockOpen(collapsed) }
+          blockType={ blockType(topBlock) }
+          blockAlign={ blockAlign(topBlock) }
+          blockRect={ rangeRect(topBlock) }
+          />
+      </div>
+      <TinyMCEReact content={window.content}
```

Comment:
# This is where it starts: 
 https://github.com/WordPress/gutenberg/pull/275/files#diff-29319b526b66e55eb72a1e1696deee28

- we create an app state 'store' here, and attach the 'actions' to it
- the actions define the state transitions (state, action) -> (state) 
- we pass the state store through to the App so it can dispatch action messages to the store
- the app subscribes to any state changes to the store

# Then the Turducken component coordinates the UI and Tiny:
https://github.com/WordPress/gutenberg/pull/275/files#diff-aa23597862e3bf21c5f904c13462266a

- it supplies state values as props to the *Toolbars (and the DIV outline blocks which we didnt have time to move out into its own component)
- it supplies dispatch functions as props to the Tiny component which dispatching action messages to the store when content state relevant to the UI changes